core/embed/trezorhal/rdi.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "memzero.h"
+#include "rand.h"
+#include "rdi.h"
+
+#define BUFFER_LENGTH 64
+#define RESEED_INTERVAL 65536
+
+static CHACHA_DRBG_CTX drbg_ctx;
+static uint8_t buffer[BUFFER_LENGTH];
+static size_t buffer_index;
+static uint8_t session_delay;
+static bool refresh_session_delay;
+static bool rdi_enabled = false;
+
+static void rdi_reseed(void) {
+  uint8_t entropy[CHACHA_DRBG_SEED_LENGTH];
+  random_buffer(entropy, CHACHA_DRBG_SEED_LENGTH);
+  chacha_drbg_reseed(&drbg_ctx, entropy);
+}
+
+static void buffer_refill(void) {
+  chacha_drbg_generate(&drbg_ctx, buffer, BUFFER_LENGTH);
+}
+
+static uint32_t random8(void) {
+  buffer_index += 1;
+  if (buffer_index >= BUFFER_LENGTH) {
+    buffer_refill();
+    if (RESEED_INTERVAL != 0 && drbg_ctx.reseed_counter > RESEED_INTERVAL)
+      rdi_reseed();
+    buffer_index = 0;
+  }
+  return buffer[buffer_index];
+}
+
+void rdi_refresh_session_delay(void) {
+  if (rdi_enabled) refresh_session_delay = true;
+}
+
+__attribute__((target("thumb"))) void rdi_handler(uint32_t uw_tick) {
```

Comment:
It might be good to add a brief comment for future generations somewhere around here, saying that this implements the floating mean distribution and maybe a link to the paper.

Reply:
I added comment to the beginning of the file.